pkgs/by-name/cm/cmake/000-nixpkgs-disable-bin-prefix.diff

```diff
@@ -0,0 +1,36 @@
+diff --git a/Source/cmSearchPath.cxx b/Source/cmSearchPath.cxx
```

Comment:
What are the chances of upstreaming this patch (possibly with modifications)?

Reply:
to upstream would want to remove env var and use a command line argument to enable/disable i think. then need an argument why users other than nixpkgs would want this. the patch as is i think works well for our purposes as it is minimal and non-intrusive. it might be worth while filing an issue upstream tho as perhaps it causes issues for other cross-compilation frameworks?

there are cmake variables  https://cmake.org/cmake/help/latest/variable/CMAKE_IGNORE_PATH.html#variable:CMAKE_IGNORE_PATH and https://cmake.org/cmake/help/latest/variable/CMAKE_IGNORE_PREFIX_PATH.html#variable:CMAKE_IGNORE_PREFIX_PATH which seem like they'd work but when set they also ignore things in PATH which doesn't work for us.

Reply:
It doesn't work in an absolute sense, or in the sense of "using this CMAKE_IGNORE_PREFIX_PATH could theoretically work, however it would require too much extra work"?

Because, well, my first thought was "well, we set this var to throw all PATH out, then we rebuild PATH as we wish".

Reply:
it does not work because it ignores the directories in PATH in addition to CMAKE_PREFIX_PATH. I think my first attempt was when we add a directory to CMAKE_PREFIX_PATH i added `dir/bin` to CMAKE_IGNORE_PATH. That has issues if the directory is also in the PATH. Then I thought i could just take the difference of CMAKE_PREFIX_PATH + bin - PATH and ignore that. But then that can change the lookup order. In the case of zig / castxml I think that the compiler wrapper needed to come first but there were still binaries that were required in the llvm directory that also contained the unwrapped compiler. so both directories were needed and preserving the order was important.

so your suggestion, which i think is to say ignore everything in CMAKE_PREFIX_PATH then build a PATH with aliases  seems like it would work. However, it would still have the issue @emilazy brought up about finding binaries outside of the bin directory, like share and lib, but i suppose those could be enumerated and just added to CMAKE_IGNORE_PATH as well.

though, i suppose we could use my previous, non cmake change, and just add `share` `lib` etc to the CMAKE_IGNORE_PATH too. I do prefer cmake patch because of it's simplicity and we don't need to enumerate all paths that cmake searches long hand and it seems exactly what we want. the downside is that we need to maintain a patch that changes cmake behavior.

Reply:
Regarding upstreaming, I get a sense that maybe cmake should need separate `CMAKE_PREFIX_PATH_HOST` and `CMAKE_PREFIX_PATH_BUILD`, with the logic that `CMAKE_PREFIX_PATH` is appended to both